// @genkit-ai/mcp/lib/util/message.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var message_exports = {};
__export(message_exports, {
  fromMcpPart: () => fromMcpPart,
  fromMcpPromptMessage: () => fromMcpPromptMessage
});
module.exports = __toCommonJS(message_exports);
const ROLE_MAP = {
  user: "user",
  assistant: "model"
};
function fromMcpPromptMessage(message) {
  return {
    role: ROLE_MAP[message.role],
    content: [fromMcpPart(message.content)]
  };
}
function fromMcpPart(part) {
  switch (part.type) {
    case "text":
      return { text: part.text };
    case "image":
      return {
        media: {
          contentType: part.mimeType,
          url: `data:${part.mimeType};base64,${part.data}`
        }
      };
    case "resource":
      return {
        resource: {
          uri: part.uri
        }
      };
    default:
      return {};
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  fromMcpPart,
  fromMcpPromptMessage
});
